---
title: "Vizes for HIV Wide"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
library(treeio)
library(tidytree)
library(ape)
library(parallel)
library(tidyverse)
library(future)
library(future.apply)
library(plotly)
#library(listenv)
plan(multisession, workers=8)
PATH="/gpfs/data/cbc/aguang/hiv_wide"
setwd(PATH)
#PATH="~/CORE/kantorlab/hiv_wide/"
all_tree_bootstraps <- readRDS(file=file.path(PATH,"all_tree_bootstraps.rds"))
```

# Overview

## Making dataframe of bootstrap values

We put the trees into a dataframe that has as columns: the percent of full genome (0 through 100), the partition (as produced by prop.part), the bootstrap value of that partition (NA if doesn't exist in tree for percent of full genome, otherwise bootstrap value)

```{r loadtree}
treefiles <- list.files(file.path(PATH,"trees"),".fa.treefile",full.names=TRUE)
trees <- lapply(treefiles, read.iqtree)
phylos <- lapply(trees, function(x) x@phylo)
bipartitions <- prop.part(phylos)
labels %<-% lapply(1:length(bipartitions), function(x) attributes(bipartitions)$labels[bipartitions[x][[1]]] %>% sort)
treenames <- c("wgs90", "wgs80", "wgs70", "wgs60", "wgs50", "wgs40", "wgs30", "wgs20", "wgs10", "pol", "wgs")
```

```{r get-bootstrap, eval=FALSE}
# inputs:
#   bipart - bipartition with labels
#   t - tree index
bootstrap_helper %<-% function(bipart, t) {
    if(is.monophyletic(phylos[[t]],bipart)) {
        mrca <- ggtree::MRCA(trees[[t]], bipart)
        boot <- dplyr::filter(trees[[t]]@data, node==mrca)$UFboot
        return(list(bipart, boot))
    }
} 

#bootstrap %<-% mclapply(labels, function(x) bootstrap_helper(x, 1))

bootstrap <- function(t) {
    future_lapply(labels, function(x) bootstrap_helper(x, t))
}

all_tree_bootstraps <- lapply(1:length(trees), bootstrap)
saveRDS(all_tree_bootstraps, file=file.path(PATH,"all_tree_bootstraps.rds"))
```

```{r, make_data_frame}
# it looks like each element in each list in all_tree_bootstraps is the same partition which is nice. So we want to save the name of the partition, and the number if it's the second value, or NA if it doesn't exist/the value is NULL

# labels is the actual list of labels

# in order to do that we have to first create a list replacement so that each element of the list has 2 elements itself
replace_null <- function(x) {
  # get number of NULL values in bootstrap list
  num_null <- sapply(x, is.null) %>% sum()
  # make list that we will use to replace, split forms them into pairs
  list_replace <- rep(list(NA,NA), num_null)
  list_replace <- split(list_replace, factor(c(1:num_null,1:num_null)))
  
  x[sapply(x,is.null)] <- list_replace
  return(x)
}

only_bootstraps <- lapply(all_tree_bootstraps, function(x)
  sapply(replace_null(x), "[[", 2))
names(only_bootstraps) <- treenames
bootstrap_df <- data.frame(only_bootstraps)
bootstrap_df$labelind <- 1:nrow(bootstrap_df)
```

We keep the labels separate from the dataframe for now because the names are extremely messy to keep as rows.

# Clustering pol phylogeny

Idea was... we are going to cluster the pol tree and annotate on that plot made of the bootstraps. I had started with getting this analysis onto OOD through a Docker image, that I remember.

```{r pol_cluster, eval=FALSE}
which(bootstrap_df$pol>99)
```

```{r pol_tree, eval=FALSE}
# spotchecking the results of this it does look ok but should still write a test tree and be careful...

poltree <- trees[[10]]
pol_bootstrap99 <- filter(poltree@data, UFboot > 99)$node
copy_pol_bootstrap99 <- pol_bootstrap99
# to cluster, for a given node with UFboot > 99, are its children also UFboot > 99?
# we can start at the beginning of the vector since those are the largest subtrees / closest to the root and delete the children from the list until we hit a leaf / there are no more children
i <- 1
node <- pol_bootstrap99[i]
tips <- 1:3860
clusters <- c()
while(!is.na(node)) {
    subnodes <- offspring(poltree@phylo, node)
    subnodes <- subnodes[-which(subnodes %in% tips)] # remove tips
    #print("subnodes")
    #print(subnodes)
    # if all subnodes are in the set of nodes with bootstrap > 99, delete the subnodes and add node
    # to set of clusters
    if (length(subnodes)==0) {
      clusters <- c(clusters, node)
    } else if (all(subnodes %in% pol_bootstrap99)) {
      copy_pol_bootstrap99 <- copy_pol_bootstrap99[-which(copy_pol_bootstrap99 %in% subnodes)]
      #print("copy_pol_bootstrap99")
      #print(copy_pol_bootstrap99)
      clusters <- c(clusters, node)
    }
    i <- i + 1
    node <- copy_pol_bootstrap99[i]
    #print(node)
}

polclust_tips <- lapply(clusters, function(x) offspring(poltree@phylo, x, tiponly=TRUE))
polclust_labels <- lapply(polclust_tips, function(x) poltree@phylo$tip.label[x] %>% sort)
polclust_index <- sapply(polclust_labels, function(x) which(sapply(labels, identical, x)))
```

Ok, now that we have the clusters that are ID'd as transmission clusters in pol with a 99% bootstrap cutoff, we want to label them on the whole genome vs proportion plot. First we want to recreate that plot using ggplot though. Basically... if a clade is in the whole alignment, plot its bootstrap value as well as any for the other trees.

Ah, but the pol tree has its own tips set...

```{r pol_v_masked}
wgsclades <- filter(bootstrap_df, !is.na(wgs)) %>%
  pivot_longer(cols=wgs90:pol, names_to="tree",values_to="bootstrap")
wgsclades$polclust <- wgsclades$labelind %in% polclust_index

num_pol_in_wgs <- sum(polclust_index %in% wgsclades$labelind)
num_pol_in_wgs
num_pol_in_wgs/length(polclust_index) # proportion

cc <- scales::seq_gradient_pal("white", "blue", "Lab")(seq(0,1,length.out=10))
ggplot(wgsclades,
       aes(x=wgs,y=bootstrap,color=tree), alpha=0.3) +
    geom_jitter(data=filter(wgsclades,labelind %in% polclust_index),
             aes(x=wgs,y=bootstrap),color="black",size=4.5,alpha=0.3) +
    geom_jitter() +
  scale_color_manual(values=cc, labels=c("pol", "wgs10-pol90", "wgs20-pol80",
                                         "wgs30-pol70","wgs40-pol60","wgs50-pol50",
                                         "wgs60-pol40","wgs70-pol30","wgs80-pol20",
                                         "wgs90-pol10")) +
  xlab("WGS Bootstrap") +
  ylab("Mixture Bootstrap") +
  ggtitle("WGS Bootstrap vs Mixture Bootstrap")
```

# Heatmap

```{r heatmap}
names(trees) <- c(90,80,70,60,50,40,30,20,10,0,100)
heatmap_df <- lapply(trees, function(x) x@data) %>% bind_rows(.id="id")
heatmap_df$id <- factor(heatmap_df$id, order=TRUE, levels=c(0,10,20,30,40,50,60,70,80,90,100))
heatmap_df <- mutate(heatmap_df, bootbins=case_when(
    UFboot < 95 ~ "<95",
    UFboot == 95 ~ "95",
    UFboot == 96 ~ "96",
    UFboot == 97 ~ "97",
    UFboot == 98 ~ "98",
    UFboot == 99 ~ "99",
    UFboot == 100 ~ "100"
))
p <- ggplot(heatmap_df, aes(x=id,y=UFboot)) + geom_bin_2d() +
    xlab("Percentage of whole genome") +
    ylab("Bootstrap values") +
    ggtitle("Number of nodes with y bootstrap value in tree \n
            with x percent of tips as whole genome sequences") +
    guides(fill=guide_legend(title="Number of nodes"))
ggplotly(p)
```